src/connector/codec/src/decoder/avro/mod.rs

```diff
@@ -266,14 +253,17 @@ impl<'a> AvroParseOptions<'a> {
                 ScalarImpl::Interval(Interval::from_month_day_usec(months, days, usecs))
             }
             // ---- Struct -----
-            (DataType::Struct(struct_type_info), Value::Record(descs)) => StructValue::new(
+            (DataType::Struct(struct_type_info), Value::Record(descs)) => StructValue::new({
+                let Schema::Record(record_schema) = &self.schema else {
+                    return Err(create_error());
+                };
```

Comment:
random thought: maybe ideally, there should be some wrapper like `ValueWithSchema::Record(descs, schema)`.

Reply:
check in each branch looks verbose and error-prone

Reply:
This annoying pattern has been common, where we match both `DataType::Struct` and `Scalar::Struct`. Are you suggesting there is a better way to handle such cases?

My own practice has been to match on `DataType` and *assert* on `Scalar` with `as_struct`. An alternative form of `as_struct` is the `try_match_expand` macro. But in this PR I tried to avoid such large structural refactor and focus on correcting the inner schema extraction within the existing gigantic match.

Btw I do not think matching them as a tuple (`(DataType, Value)` or `(DataType, Value, Schema)`) is good either. It means we inevitably need a wildcard `_` and compiler would not remind us of unhandled data types.

Reply:
It's like this:
Basically wrap your practice into a separate step, and later we can just exhaustively match.
```rust
enum ScalarWithType {
    Struct(struct_value, struct_type),
    ....
}

impl Scalar {
    fn combine_type(self, type: DataType) -> ScalarWithType {
        match type {
            DataType::Struct(struct_type) => { 
                ScalarWithType(self.as_struct(), struct_type)
            }
            ...
        }
    }
}
```